### PR TITLE
Improve method name suggestions

### LINKED ISSUE #94747: improve similar method names suggestions
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af286f54d571ef5e00cd728ea244b3aa

```rust
fn main() {
    vec![].bs_xmpty();
    vec![].empty();
}
```

The current output is:

```
error[[E0599]](https://doc.rust-lang.org/stable/error-index.html#E0599): no method named `bs_xmpty` found for struct `Vec<_>` in the current scope
 [--> src/main.rs:2:12
](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af286f54d571ef5e00cd728ea244b3aa#)  |
2 |     vec![].bs_xmpty();
  |            ^^^^^^^^ help: there is an associated function with a similar name: `is_empty`

error[[E0599]](https://doc.rust-lang.org/stable/error-index.html#E0599): no method named `empty` found for struct `Vec<_>` in the current scope
 [--> src/main.rs:3:12
](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af286f54d571ef5e00cd728ea244b3aa#)  |
3 |     vec![].empty();
  |            ^^^^^ method not found in `Vec<_>`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to 2 previous errors
```

Ideally the output should look like:

```
error[[E0599]](https://doc.rust-lang.org/stable/error-index.html#E0599): no method named `bs_xmpty` found for struct `Vec<_>` in the current scope
 [--> src/main.rs:2:12
](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af286f54d571ef5e00cd728ea244b3aa#)  |
2 |     vec![].bs_xmpty();
  |            ^^^^^^^^ help: there is an associated function with a similar name: `is_empty`

error[[E0599]](https://doc.rust-lang.org/stable/error-index.html#E0599): no method named `empty` found for struct `Vec<_>` in the current scope
 [--> src/main.rs:3:12
](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af286f54d571ef5e00cd728ea244b3aa#)  |
3 |     vec![].empty();
  |            ^^^^^^^^ help: there is an associated function with a similar name: `is_empty`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to 2 previous errors
```

To a human, the method name `is_empty` is much closer to `empty` than to `bs_xmpty`.
However, the algorithm used by the compiler to suggest similar method names seems a little bit simplistic and fails to suggest anything when a naive programmer like me types `empty()` instead of `is_empty`. The programmer then has to open the documentation, and he does not like that ;)

### PR DESCRIPTION
**Pull request by OliverMD**
Attempts to improve method name suggestions when a matching method name
is not found. The approach taken is use the Levenshtein distance and
account for substrings having a high distance but can sometimes be very
close to the intended method (eg. empty vs is_empty).

resolves #94747

### CODE REVIEW

File: compiler/rustc_span/src/lev_distance.rs [hunk 1]
```diff
@@ -46,6 +46,49 @@ pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
...
 }
 
+/// Provides a method name similarity score between two method names. The lower the score,
+/// the closer the match. 0 is an identical match.
+///
+/// Uses the Levenshtein distance between the two strings and attempts to account
+/// for cases where one is a substring of the other.
+///
+/// Returns None if the distance exceeds the limit
+pub fn method_name_similarity_score(a: &str, b: &str, limit: usize) -> Option<usize> {
```
Comment (davidtwco) [inline compiler/rustc_span/src/lev_distance.rs:56]: We could probably give this a more generic name, `lev_distance_with_substrings` or something like that.

File: compiler/rustc_span/src/lev_distance.rs [hunk 2]
```diff
@@ -46,6 +46,49 @@ pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
...
+    let m = b.chars().count();
+
+    // Check one isn't less than half the length of the other
+    let big_len_diff = (n * 2) < m || (m * 2) < n;
+    let len_diff = if n < m { m - n } else { n - m };
+    let lev = lev_distance(a, b, limit + len_diff)?;
+
+    let score = lev - len_diff;
+
+    let score = {
```
Comment (davidtwco) [inline compiler/rustc_span/src/lev_distance.rs:67]: nit: the outer braces are unnecessary

File: compiler/rustc_span/src/lev_distance.rs [hunk 3]
```diff
@@ -46,6 +46,49 @@ pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
...
+    };
+
+    (score <= limit).then_some(score)
+}
+
+/// Finds the best match for given method name in the given iterator.
+///
+/// A version of [`find_best_match_for_name`] that uses a modified Levenshtein score
+/// to provide more useful matches for method names.
+pub fn find_best_match_for_method_name(
```
Comment (davidtwco) [inline compiler/rustc_span/src/lev_distance.rs:84]: As before, I think these would be better if they didn't mention method names. It's not obvious what modifications to a Levenshtein score are necessary for method names when this is the only comment you've seen, so it's probably better to describe how it is modified.

File: compiler/rustc_span/src/lev_distance.rs [hunk 4]
```diff
@@ -46,6 +46,49 @@ pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
...
+
+    // Check one isn't less than half the length of the other
+    let big_len_diff = (n * 2) < m || (m * 2) < n;
+    let len_diff = if n < m { m - n } else { n - m };
+    let lev = lev_distance(a, b, limit + len_diff)?;
+
+    let score = lev - len_diff;
+
+    let score = {
+        if score == 0 && len_diff > 0 && !big_len_diff {
```
Comment (davidtwco) [inline compiler/rustc_span/src/lev_distance.rs:68]: Some comments might be helpful to explain the rationale behind these conditionals - they make sense if I think long enough but it's easier when I don't need to :)

File: compiler/rustc_span/src/lev_distance.rs [hunk 5]
```diff
@@ -46,6 +46,49 @@ pub fn lev_distance(a: &str, b: &str, limit: usize) -> Option<usize> {
     (dcol[m] <= limit).then_some(dcol[m])
 }
 
+/// Provides a method name similarity score between two method names. The lower the score,
+/// the closer the match. 0 is an identical match.
+///
+/// Uses the Levenshtein distance between the two strings and attempts to account
+/// for cases where one is a substring of the other.
+///
+/// Returns None if the distance exceeds the limit
```
Comment (davidtwco) [inline compiler/rustc_span/src/lev_distance.rs:55]:
```suggestion
/// Returns `None` if the distance exceeds the limit.
```
nit

File: src/test/ui/derives/issue-91550.stderr [hunk 6]
```diff
@@ -8,7 +8,7 @@ LL | struct Value(u32);
    | doesn't satisfy `Value: Hash`
 ...
 LL |     hs.insert(Value(0));
-   |        ^^^^^^ method cannot be called on `HashSet<Value>` due to unsatisfied trait bounds
```
Comment (davidtwco) [inline src/test/ui/derives/issue-91550.stderr:11]: I'm not sure that this is necessarily an improvement, it doesn't hurt the quality of the error dramatically or anything like that but it's probably a little worse-off.

**Review by davidtwco — CHANGES_REQUESTED** [review]
This is a good improvement, apologies that it took a while for this to be reviewed, I've left a few comments.

File: src/test/ui/derives/issue-91550.stderr (hunk 6 above)
Comment (OliverMD) [reply src/test/ui/derives/issue-91550.stderr:11]: Yeah, I think this is one of the examples where the quality is slightly worse, one the challenges with this has been balancing the consequences of the change. Do you think this is a blocker or do the benefits of the change outweigh this?

File: compiler/rustc_span/src/lev_distance.rs (hunk 5 above)
Comment (OliverMD) [reply compiler/rustc_span/src/lev_distance.rs]: Fixed

File: compiler/rustc_span/src/lev_distance.rs (hunk 4 above)
Comment (OliverMD) [reply compiler/rustc_span/src/lev_distance.rs]: Added some comments around this and gave a bit more context for what we're trying to do in the rustdoc

File: compiler/rustc_span/src/lev_distance.rs (hunk 3 above)
Comment (OliverMD) [reply compiler/rustc_span/src/lev_distance.rs]: I think I've removed all the references to method now

File: compiler/rustc_span/src/lev_distance.rs (hunk 2 above)
Comment (OliverMD) [reply compiler/rustc_span/src/lev_distance.rs]: Fixed

File: compiler/rustc_span/src/lev_distance.rs (hunk 1 above)
Comment (OliverMD) [reply compiler/rustc_span/src/lev_distance.rs]: Changed to `lev_distance_with_substrings`

**Comment by OliverMD** [comment]
Thanks for the review @davidtwco! I've pushed some changes based on the review.

File: src/test/ui/derives/issue-91550.stderr (hunk 6 above)
Comment (davidtwco) [reply src/test/ui/derives/issue-91550.stderr:11]: I don't think this is a blocker, in each of the tests that have changed like this, all of the information that was lost is available elsewhere in the diagnostic.

**Review by davidtwco — APPROVED** [review]
Thanks, this is a good improvement!

**Pull request merged**